ClusterSearcher and RecoCluster

flemmons opened the pull request:
-Updated DigitBuilder to be more consistent with and independent of ClusterFinder's corresponding functions
--Added some extra features, such as hit LAPPD count and strip-by-strip LAPPD hits. (some of these are considered experimental/preliminary)
-Added ClusterSearcher to replace ClusterFinder using RecoDigit and RecoCluster classes 
-Updated RecoDigit and RecoCluster classes for corresponding use and new features, such as various cluster parameters 
-Added NeutronCheck tool as output for RecoCluster information 
-Added sample toolchain configfolder for using the new tools

flemmons committed:
-Updated DigitBuilder to be more consistent with, but independent of ClusterFinder's corresponding functions
--Added some extra features, such as hit LAPPD count and strip-by-strip LAPPD hits.
-Added ClusterSearcher to replace ClusterFinder using RecoDigit and RecoCluster classes
-Updated RecoDigit and RecoCluster classes for corresponding use and new features, such as various cluster parameters
-Added NeutronCheck tool as output for RecoCluster information
-Added sample toolchain configfolder for using the new tools

jminock commented on UserTools/DigitBuilder/DigitBuilder.cpp:
```diff
@@ -66,37 +73,66 @@ bool DigitBuilder::Initialise(std::string configfile, DataModel &data){
+    }
+
+
+    /*while (!file_singlepe.eof()) {
```
Please delete the old commented out code. It unnecessarily clutters the file.

jminock commented on DataModel/RecoCluster.cpp:
```diff
@@ -48,5 +48,440 @@ int RecoCluster::GetNDigits()
+            //}
+        }
+    }
+    //FIXME: Need a method to have the 123 be equal to the number of operating detectors
```
Should this be fixed now?

jminock commented on DataModel/RecoCluster.cpp:
```diff
@@ -48,5 +48,440 @@ int RecoCluster::GetNDigits()
+        double max_angle = 0;
+        double angle;
+        Position i_position, j_position;
+        for (int i = 0; i < fDigitList.size(); i++) {
```
There are many separate loops of the same size with the same index going on. Could we combine them? If this is a DataModel function, it could be used across multiple Tools so it would be best to improve efficiency within reason.

S81D commented on configfiles/ClusterSearcher/LoadWCSimConfig:
```diff
@@ -0,0 +1,17 @@
+LappdNumStrips 60            ## num channels to construct from each LAPPD
+LappdStripLength 100         ## relative x position of each LAPPD strip, for dual-sided readout [mm]
+LappdStripSeparation 10      ## stripline separation, for calculating relative y position of each LAPPD strip [mm]
+PMTMask configfiles/BeamClusterAnalysisMC/DeadPMTIDs_p2v7.txt ## Which PMTs should be masked out? / are dead?
```
You can point to the most up to date path `configfiles/LoadWCSim/DeadPMTIDs_p2v7.txt`

S81D commented on UserTools/ClusterSearcher/ClusterSearcher.cpp:
```diff
@@ -0,0 +1,827 @@
+    if (!fisMC){
+      int pmtid = recoDigit->GetDetectorID();
+      unsigned long chankey = pmt_tubeid_to_channelkey[pmtid];
+      if (pmt_gains[chankey]>0) qep/=pmt_gains[chankey];
```
In line 77 (`m_variables.Get("SinglePEGains",singlePEgains);`) you grab the SPE gains from the gains file then use it to convert from nQ --> pe for data. It doesn't look like you ever specify the gains file in the Config file. Could you add it to `ClusterSearcherConfig`? Alternatively you can grab that directly from the Store since the `LoadGeometry` tool populates it for downstream tools (like this one) to use. See `PhaseIITreeMaker` for an example.

flemmons replied:
The current configuration in the provided toolchain is for use on MC simulations, which do not require the conversion.  So this variable is not necessary within the configuration.  I will add it to the readme file.
Though as I start looking at Data in the next couple of weeks, I'll likely move into taking the gains from the store in the next update to this tool.  Thank you for the suggestion.

S81D commented on UserTools/NeutronCheck/NeutronCheck.cpp:
```diff
@@ -0,0 +1,491 @@
+            if(fMCParticles->at(i).GetPdgCode()==2112 && fMCParticles->at(i).GetParentPdg()==0) {
+                fTrueNeutronMult++;
+                
+                if(fMCParticles->at(i).GetStopTime()>10000) fTrueNeutronDelayed++;
```
It might be wise to make this a configurable variable in case someone needs to look for neutrons in a different region of interest. Especially considering 10us is used extensively throughout the code.

S81D commented on UserTools/NeutronCheck/NeutronCheck.cpp:
```diff
@@ -0,0 +1,491 @@
+        //true_Emu*=1000;  //GeV->MeV to match other energies(unneeded, possibly)
+
+        double theta = truevtx->GetDirection().GetTheta();
+        double p = sqrt(pow(true_Emu,2)-pow(105.7,2));
```
@jminock is there a way to grab the momentum and Q^2 from the Store? I thought the `LoadGenieEvent` tool will store those values for use.

jminock replied:
There absolutely is via: `m_data->Stores["GenieInfo"]->Get("EventQ2",TrueQ2)`. Magnitude of muon momentum is not saved to the Store so this is the intended method to get the true muon momentum

marc1uk commented on DataModel/RecoCluster.cpp:
```diff
@@ -48,5 +48,440 @@ int RecoCluster::GetNDigits()
+    for (RecoDigit* i_digit : fDigitList) {
+        cout<<"Digit parent list size: "<<i_digit->GetParents().size()<<endl;
+        if (i_digit->GetParents().size() == 0) {
+            cout<<"Digit found with no parents!!!\n";
```
Is this sufficient error-handling behaviour? It seems this will return 0 as the best parent index - perhaps this is fine, or perhaps it should return an invalid index, or throw an exception, or something else....I'm not familiar enough with the code to know, so just checking.

flemmons replied:
Since PDG 0 is unused within ToolAnalysis, it ultimately works as an error flag.  It's used by photons in Genie, as I understand it, but those are not passed to TA as particles, so the index is invalid in our context.  And it's the only convenient 'invalid' index I'm aware of (-9999, the usual go-to, is likely to be an anti-matter nucleus of some variety, which we won't see in our experiment, but I don't think it's technically invalid).  So I'll argue that this is correct.

marc1uk replied:
ok... i'm not sure how that response quite ties up with the question, maybe I was unclear. 
This is the `calcBestParent` function, and how it handles digits with no parents. Right now it just prints to screen and then carries on. Not sure what that has to do with PDG 0.
Anyhow, with a fresh pair of eyes I see the behaviour here is to ignore the charge contribution from any digits without parents when calculating the best parent particle for the cluster, which does seem like the only appropriate thing one can do. So i'll resolve this comment.

marc1uk commented on DataModel/RecoCluster.cpp:
```diff
@@ -48,5 +48,440 @@ int RecoCluster::GetNDigits()
+    double sumY=0;
+    Position pos;
+    //std::vector<RecoDigits> hullDigits;
+    std::sort(fDigitList.begin(), fDigitList.end());
```
is sorting a vector of pointers with the default comparator going to do what you expect? I see you have a `operator<` for RecoDigit, but think you still need to provide a comparitor lambda to perform the pointer de-reference:
```
std::sort(fDigitList.begin(), fDigitList.end(), [](const RecoDigit* a, const RecoDigit* b){ return a < b; });
```

flemmons replied:
I'm... just going to remove the RecoCluster::ConvexHull function, at least for now.  It's not actually useful right now.

jminock commented on DataModel/RecoCluster.cpp:
```diff
@@ -48,5 +48,440 @@ int RecoCluster::GetNDigits()
+    double maxCharge=0;
+    int tempPDG = -5;
+    for (RecoDigit* i_digit : fDigitList) {
+        cout<<"Digit parent list size: "<<i_digit->GetParents().size()<<endl;
```
Please add verbosity conditions to the cout's throughout this file and others. Preferably Log functions if you would like Marcus to give approval down the road

jminock commented on UserTools/ClusterSearcher/ClusterSearcher.cpp:
```diff
@@ -0,0 +1,827 @@
+  
+  // Default Clustering parameters
+  fConfig = ClusterSearcher::kPulseHeightAndClusters;
+  fPmtMinPulseHeight = 5.0;     // minimum pulse height (PEs) //Ioana... initial 1.0
```
The configuration file should take care of the initialization. Leave the default parameters in the configuration file. If any of these NEED to be initialized outside of the configuration file, please do so in the header file upon declaration. It cuts down on clutter

jminock commented on UserTools/ClusterSearcher/ClusterSearcher.cpp:
```diff
@@ -0,0 +1,827 @@
+    double temp_gain;
+    while (!file_singlepe.eof()){
+      file_singlepe >> temp_chankey >> temp_gain;
+      if (file_singlepe.eof()) break;
```
This line seems unnecessary. While loop already includes this condition

marc1uk replied:
i think it would also prevent the `emplace` call for the last line of the file.

jminock commented on UserTools/ClusterSearcher/ClusterSearcher.cpp:
```diff
@@ -0,0 +1,827 @@
+
+      carryon = 1;
+      while( carryon ){
+        carryon = 0;
```
Why is this for loop inside of a while loop?

flemmons replied:
Seemingly because only one of the two loops had a well-defined range over which to run.  This code was inherited from the HitCleaner tool, so I can't answer the why of that decision with complete certainty, but the code flows to my reading, and works as needed.
This seems like a pet peeve, more than a problem, so for the time being, I'll stand by it as is.

jminock commented on UserTools/DigitBuilder/DigitBuilder.cpp:
```diff
@@ -18,7 +18,7 @@ DigitBuilder::~DigitBuilder() {
 
   /////////////////// Usefull header ///////////////////////
-  if(verbosity) cout<<"Initializing Tool DigitBuilder"<<endl;
+  cout<<"Initializing Tool DigitBuilder"<<endl;
```
I think this has been mentioned before. Please include verbosity conditions, preferably as Log functions

jminock reviewed (changes requested):
Please make the changes listed. I worry with the large number of pointers if memory is being handled properly. I don't know how necessary all of the pointers are, and I am not enough of an expert on it to make a certain statement. I recommend double checking all are fitting general best practice before this gets sent off to Level 0 review

marc1uk commented on UserTools/ClusterSearcher/ClusterSearcher.cpp:
```diff
@@ -0,0 +1,827 @@
+  }
+
+  if( !fgClusterSearcher ){
+    assert(fgClusterSearcher);
```
this seems redundant. If the previous allocation didn't succeed it would have thrown `bad_alloc`.

marc1uk commented on UserTools/ClusterSearcher/ClusterSearcher.cpp:
```diff
@@ -0,0 +1,827 @@
+  }
+
+  if( fgClusterSearcher ){
+
```
⠀

marc1uk commented on UserTools/ClusterSearcher/ClusterSearcher.cpp:
```diff
@@ -0,0 +1,827 @@
+  }
+
+  if (!fisMC){
+    ifstream file_singlepe(singlePEgains.c_str());
```
recommend adding a check `file_singlepe.is_open()` to catch typos in filename.

marc1uk commented on UserTools/ClusterSearcher/ClusterSearcher.cpp:
```diff
@@ -0,0 +1,827 @@
+    return false;
+  }
+  // copy to a new digit list
+  std::vector<RecoDigit*>* digits = new std::vector<RecoDigit*>;
```
does this need to be on the heap?

marc1uk commented on UserTools/ClusterSearcher/ClusterSearcher.cpp:
```diff
@@ -0,0 +1,827 @@
+  // Select all digits
+  // =================
+  InputList = ResetDigits(OutputList); 
+  OutputList = (std::vector<RecoDigit*>*)(this->SelectAll(InputList));
```
the cast seems unnecessary: `ClusterSearcher::SelectAll` is already defined to return `std::vector<RecoDigit*>*`

n.b. use of `this` at various points is also unnecessary.

marc1uk commented on UserTools/ClusterSearcher/ClusterSearcher.cpp:
```diff
@@ -0,0 +1,827 @@
+  InputList = ResetDigits(OutputList);
+  OutputList = (std::vector<RecoDigit*>*)(this->SelectByPulseHeight(InputList));
+  OutputList = SelectDigits(OutputList);
+  if( fConfig==ClusterSearcher::kPulseHeight ) return OutputList;
```
it seems the pattern here is
1. reset filter status
2. remove some digits
3. set filter status

Maybe i'm missing it, but it looks to me as though the `SelectBy*` functions don't actually use the filter status. So wherever this code returns, it will return a list of digits all of which have `fIsFiltered` set, and the repeated `ResetDigits`/`SelectDigits` calls are unnecessary.

marc1uk commented on UserTools/ClusterSearcher/ClusterSearcher.cpp:
```diff
@@ -0,0 +1,827 @@
+  delete fSelectByPulseHeight; fSelectByPulseHeight = 0;
+  delete fSelectByNeighbours; fSelectByNeighbours = 0;
+  delete fSelectByClusters; fSelectByClusters = 0;
+  //delete fRecoClusters; fRecoClusters = 0;    //Will be deleted by the store, don't manually delete
```
As far as I can see, `fRecoClusters` is not actually put into any CStore, so it will not be deleted by the store. Please uncomment this, unless i'm mistaken.

More importantly, though, is that the `fRecoClusters` holds pointers to RecoCluster objects that are also created by this Tool on the heap. Those pointers get transferred into the "RecoEvent" store "RecoClusters" entry.... but that makes it very difficult to tell whether those RecoCluster objects are deleted. The CStore won't take care of deleting heap-allocated objects within containers. Please double check that those objects are being deleted somewhere.

marc1uk commented on UserTools/ClusterSearcher/ClusterSearcher.cpp:
```diff
@@ -0,0 +1,827 @@
+
+  // run clustering algorithm
+  // ========================
+  std::vector<RecoCluster*>* ClusterList = (std::vector<RecoCluster*>*)(this->RecoClusters(DigitList));
```
again shouldn't need the c-style cast - in fact the local variable ClusterList is just an alias for the member variable fRecoClusters, so seems redundant. The aliasing in this Tool makes it hard to track objects.

marc1uk commented on UserTools/ClusterSearcher/ClusterSearcher.cpp:
```diff
@@ -0,0 +1,827 @@
+  // ===================
+  for(int idigit=0; idigit<int(fSelectByNeighbours->size()); idigit++ ){
+    RecoDigit* recoDigit = (RecoDigit*)(fSelectByNeighbours->at(idigit));
+    RecoClusterDigit* clusterDigit = new RecoClusterDigit(recoDigit);
```
why do the digits need to be on the heap?

marc1uk commented on UserTools/DigitBuilder/DigitBuilder.cpp:
```diff
@@ -66,37 +73,66 @@ bool DigitBuilder::Initialise(std::string configfile, DataModel &data){
+        //Loop over lines, collect all detector data (should only be one line here)
+        while (getline(file_singlepe, line)) {
+            if (verbosity > 3) std::cout << line << std::endl; //has our stuff;
+            if (line.find("#") != std::string::npos) continue;
```
be wary that this will skip lines with trailing comments, as well as commented-out lines.
perhaps `if(line.empty() || line[0]=='#') continue` is a safer/clearer check

marc1uk commented on UserTools/DigitBuilder/DigitBuilder.cpp:
```diff
@@ -66,37 +73,66 @@ bool DigitBuilder::Initialise(std::string configfile, DataModel &data){
+            if (verbosity > 3) std::cout << line << std::endl; //has our stuff;
+            if (line.find("#") != std::string::npos) continue;
+            std::vector<std::string> DataEntries;
+            boost::split(DataEntries, line, boost::is_any_of(","), boost::token_compress_on);
```
i think `token_compress_on` merges repeated tokens? Is this desirable? If there are repeated tokens, doesn't this suggest an empty column, and by compressing that, your later columns will be shifted?

marc1uk commented on UserTools/DigitBuilder/DigitBuilder.cpp:
```diff
@@ -243,69 +293,88 @@ bool DigitBuilder::BuildMCPMTRecoDigit() {
+              Log("This HIT'S TIME AND CHARGE: " + to_string(ahit.GetTime()) + ", " + to_string(ahit.GetCharge()),v_debug,verbosity);
             double hitTime = ahit.GetTime()*1.0;
-          	if(hitTime>-10 && hitTime<40) {
+          	if(hitTime>-10 && hitTime<70) {
```
better for magic numbers to be configuration variables, especially if they are subject to change

flemmons committed:
Update to previous commit in accordance with feedback:
-Changed RecoCluster and RecoDigit lists in several tools to be vectors of objects, rather than vectors of pointers, to avoid memory complications
-removed unused convex hull function from RecoCluster class
-simplified CalcAS function in RecoCluster class by consolidating for loops
-removed several instances of commented-out code from older versions
-removed several debug outputs
-altered hard-coded time window values in several instances to rely on configuration input
-added use of the true Q2 value from the GenieInfo store to NeutronCheck's output.
-removed several uncontrolled cout lines, and replaced useful ones with Log.
-Tidied the Instance() function of ClusterSearcher
-Added vertex information to NeutronCheck

jminock commented:
Thank you @flemmons for the updates! Unfortunately, this branch has conflicts with the main branch that must be resolved in order to be merged. I will wait for the conflicts to be resolved and for the workflow check to be performed such that ToolAnalysis is confirmed to compile before I review the actual changes to the files

flemmons commented:
Conflicts were just Factory and Unity listing different added tools.  I've cleared them.

jminock commented:
Okay, but could you still resolve those conflicts? Also, it's great that ToolAnalysis can compile in your workspace, but it would be better to get confirmation that ToolAnalysis can compile in a general workspace (on GitHub via workflow). Given that this PR is self contained, there is no reason to not resolve these conflicts. I will look at the other files in the meantime.

flemmons committed:
Merge branch 'Application' into Application

flemmons committed:
Removed ClusterCA from NeutronCheck to fix bug.

flemmons committed:
Modified HitCleaner for non-pointer Digits list.

flemmons committed:
Update/fix to ClusterSearcher sample toolchain.

jminock commented on configfiles/ClusterSearcher/EventSelectorConfig:
```diff
@@ -0,0 +1,18 @@
+MRDRecoCut 0
+RecoPMTVolCut 0
+RecoFVCut 0 
+ArgonFV 0
```
Not a hold up for merger, but what is ArgonFV? That isn't expected for current the current EventSelector

flemmons replied:
ArgonFV is a holdover from a previous experimental study.  It...shouldn't be there.  Good catch.

jminock commented on configfiles/ClusterSearcher/LoadGenieEventConfig:
```diff
@@ -0,0 +1,12 @@
+verbosity 0
+FluxVersion 0  # use 0 to load genie files based on bnb_annie_0000.root etc files
```
The flux files corresponding to version 0 are very outdated. None of the GENIE files I produced use those form of flux files. I don't even know where those flux files are. Pointing this out more as a courtesy for your consideration

flemmons replied:
That's good to know...  And likely explains why I've seen strange behavior in the extracted true q2 values I've been working with since adding them from LoadGenieEvent.  I appreciate it.

jminock commented on configfiles/ClusterSearcher/ToolsConfig:
```diff
@@ -0,0 +1,11 @@
+MCParticleProperties MCParticleProperties ./configfiles/ClusterSearcher/MCParticlePropertiesConfig
+DigitBuilder DigitBuilder ./configfiles/ClusterSearcher/DigitBuilderConfig
+ClusterSearcher ClusterSearcher ./configfiles/ClusterSearcher/ClusterSearcherConfig
+ClusterSearcher2 ClusterSearcher ./configfiles/ClusterSearcher/ClusterSearcher2Config
```
You have to use ClusterSearcher twice? Would a second run of ClusterSearcher overwrite the first use? Sorry, just confused why this is here. Also, how does the information get written out to a file? Does NeutronCheck double as an output file maker? Tools are meant to be singular in purpose, and there already exists output file makers (PhaseIITreeMaker)

flemmons replied:
ClusterSearcher is designed to be used multiple times, yes.  It will add the new clusters to the prior list.  This enables  different sorts of clusters to be made for muli-level analysis.  The current setup is running ClusterSearcher to make concentrated Cherenkov ring-like clusters to analyze the initial muon emission for vertexing and CC verification, and then ClusterSearcher2 collects other clusters without any spatial density requirements to be tested for neutron identification.  The different types of clusters are separated by the ClusterMode tag.

As to NeutronCheck, it was meant to be more of a debug tool, akin to VertexGeometryCheck but for the neutron analysis.  Its scope grew as I needed more relevant values, but it still doesn't represent a final-state of analysis.  In the current ToolAnalysis, it's the only 'end of toolchain' tool that uses RecoClusters, and so adding it was the best option for this toolchain.  However, I'm also not sure that PhaseIITreeMaker should be the be-all, end-all output tool for all of ANNIE.  Having all output centered in a single tool will make it very cumbersome for focused analyses that only need certain variables.  It may be worth considering having separate output tools for separate analyses.  As I finalize the neutron multiplicity analysis, there will likely be a more concrete output, whether that's adding the important variables to PhaseIITreeMaker, advancing NeutronCheck to a more immediately useful form, or something else.

jminock replied:
Gotcha, please document this and how to use ClusterSearcher in a central, accessible place.

As far as the opinion on separate output Tools: having multiple Tools with similar functionality that could be altered via a configuration file is against the intended direction of ANNIE analysis and software. Different sets of Tools for different analyses will have duplicated work, divergent branches, and different standards for different analyses. Yes, PhaseIITreeMaker is cumbersome to work with. But I can guarentee standardizing different Tools with the sane functionality is much more cumbersome. There already is ANNIETreeMaker as another output file maker. I'm not opposed to that Tool being the new standard, but there should be a singular standard across analyses. Making a new separate output Tool will result in: it continuing to advance and develop after you, diverging from other analysis ToolChains, silo'ing off neutron analysis from everything else; or it getting scrapped in favor of shared output in a Tool, losing important work done now, making unnecessary work in the future. Do what you need for your present analysis, but please keep in mind that existing resources are to here to help each of us and the work we do serves ourselves, each other,  and those who come after.

flemmons replied:
That is detailed in the ClusterSearcher tool's readme file.

You're right.  I'll plan to fold the final analysis outputs into PhaseIITreeMaker or the like as I complete them, and hold NeutronCheck to the level of a debugging check of mid-level values.  But, in that role, and for the time being, it is useful, and should not hold back this pull request.  Take it as a demonstration of what ClusterSearcher is doing, more than any analysis output.

jminock commented on UserTools/HitCleaner/HitCleaner.cpp:
```diff
@@ -738,3 +740,32 @@ std::vector<RecoDigit*>* HitCleaner::FilterByTruthInfo(std::vector<RecoDigit*>*
+        //}
+    }
+    //FIXME: Need a method to have the 123 be equal to the number of operating detectors
+    double ucharge_balance = sqrt((total_QSquared) / (total_Q * total_Q) - (1. / 123.));
```
123 is a magic number. While it doesn't impact functionality for now, it can be troublesome for documentation and updating it. What does 123 represent? What is an "operating detector"?

jminock commented on UserTools/HitCleaner/HitCleaner.cpp:
```diff
@@ -523,14 +524,15 @@ std::vector<RecoDigit*>* HitCleaner::FilterByClusters(std::vector<RecoDigit*>* m
-    for(int idigit=0; idigit<myCluster->GetNDigits(); idigit++ ){
-      RecoDigit* myDigit = (RecoDigit*)(myCluster->GetDigit(idigit));
+    for(int idigit=0; idigit<myCluster.GetNDigits(); idigit++ ){
+      RecoDigit* myDigit = new RecoDigit;
```
You are creating pointers in a nested loop. How are you freeing the memory once they're done? This looks like it can and will cause memory issues.

flemmons replied:
I've removed the FilterByClusters function from use.  Most of HitCleaner is probably not working after the digit lists were changed to not include pointers.  It's structured with the assumption that all the lists are pointing to the same objects, which can no longer be the case.  I may look into fixing it for the vertex reconstruction PR which I plan to submit soon.

jminock commented on UserTools/NeutronCheck/NeutronCheck.cpp:
Why does this Tool exist when there already exists output file maker Tools?

flemmons replied:
As to NeutronCheck, it was meant to be more of a debug tool, akin to VertexGeometryCheck but for the neutron analysis. Its scope grew as I needed more relevant values, but it still doesn't represent a final-state of analysis. In the current ToolAnalysis, it's the only 'end of toolchain' tool that uses RecoClusters, and so adding it was the best option for this toolchain. However, I'm also not sure that PhaseIITreeMaker should be the be-all, end-all output tool for all of ANNIE. Having all output centered in a single tool will make it very cumbersome for focused analyses that only need certain variables. It may be worth considering having separate output tools for separate analyses. As I finalize the neutron multiplicity analysis, there will likely be a more concrete output, whether that's adding the important variables to PhaseIITreeMaker, advancing NeutronCheck to a more immediately useful form, or something else.

jminock reviewed (commented):
Please address all the comments. I left some that are more general questions that would be nice to have accessible documentation addressing, and shouldn't be an issue to create. There are some comments that ask about memory. Please address them as they do impact functionality

flemmons committed:
Removed HitCleaner::FilterByClusters because it is neither functional nor memory-safe with recent changes to RecoDigit storage.

jminock reviewed (approved):
Enough comments have been addressed such that any remaining comments are specific to this analysis. I think this is in good enough shape to merge so other neutron analyzers have access to these Tools to further their development.

marc1uk commented:
It would be helpful to 'resolve' those comments that have been addressed to make it clearer for reviewers what is still outstanding, without having to go back through the code again. 

I'm not sure I agree that the remaining issues are all analysis-specific. 
* [This comment](https://github.com/ANNIEsoft/ToolAnalysis/pull/365#discussion_r2780400680) suggests that the toolchain won't be writing out correct true q2 values, which seems important to me and could very well be difficult to track down if you're not aware of the reason. 
* `RecoCluster::CalcCB()` still seems to have a placeholder value of `123` with a comment `//FIXME: Need a method to have the 123 be equal to the number of operating detectors` - that should be simple enough to retrieve from the geometry (`ANNIEGeometry::GetNumPMTs` or similar). `CalcCB` is called by `CalcParameters` which is called by the ClusterSearcher Tool, so it seems like this information is not an irrelevant side-effect if incorrect. 
  * (The same issue is also in `HitCleaner::CBCheck`, and while that seems less important as all it does is print things, my feeling is that if the code isn't functional it shouldn't be merged, so either delete CBCheck or correct it.)
* [this comment](https://github.com/ANNIEsoft/ToolAnalysis/pull/365#discussion_r2523293347) as mentioned seems like a bug that would omit a channel. I haven't run the code to check that, but nor is there any comment to say it doesn't. If it does, this is a bug I feel is worth fixing before merge.
*  I know we're about to merge Yue's code that has several memory leaks but we shouldn't make a habit of it. Please address [this one](https://github.com/ANNIEsoft/ToolAnalysis/pull/365/changes#r2524331290).

marc1uk commented:
regarding [this discussion](https://github.com/ANNIEsoft/ToolAnalysis/pull/365/changes#r2780189059) on the output file from NeutronCheck, I'm inclined to agree with Franklin. Generating debug files is no bad thing - it allows focused validation and debugging, without polluting general-purpose analysis outputs. Unless these variables are likely to be useful for analysis, I would keep them where they are.
In general, I also agree that the PhaseIITreeMaker is not a good general-purpose analysis writer. It was originally a hack Tool to allow quick generation of ROOT plots. It now makes a tree with potentially **334** branches and counting. That's absurd! It has to be updated in-sync with every other Tool that adds or modifies what goes into it. The Tool, and its output files, will only grow as more Tools add to it. I don't personally consider it maintainable, or an appropriate way to work.
At minimum we could put a Tree in the DataModel and allow Tools to add branches to it, and have a simple Writer tool at the end of the Toolchain that calls `TTree::Fill`. That way at least each Tool remains responsible for its content, and the output file reflects what Tools were in the ToolChain, keeping content relevant to the analysis as per Franklin's request.
But of course the ANNIEEvent BoostStore is intended to be ANNIE's official output, and I'm really not sure why there's such an avoidance of BoostStores in favour of the `PhaseIIColossalTreeMaker` output instead....

marc1uk commented:
I don't mean to add further issues, and this is probably not the right place for it, but I can't help but notice that:
* ClusterSearcherConfig
* DigitBuilderConfig
* MCRecoEventLoaderConfig

all have a config variable `isMC`. While setting 3 variables isn't an especially big burden, it would seem more sensible that the appropriate [data reader Tool](https://github.com/ANNIEsoft/ToolAnalysis/blob/Application/UserTools/LoadRATPAC/LoadRATPAC.cpp#L267) sets [a flag in the DataModel](https://github.com/ANNIEsoft/ToolAnalysis/blob/Application/UserTools/LoadWCSim/LoadWCSim.cpp#L633) and Tools check that, rather than requiring multiple flags be appropriately set for the given input files.

flemmons committed:
-removed unused configvariable ArgonFV from ClusterSearcher/EventSelectorConfig
-removed unhelpful forced loop end from single-pe gains extraction in ClusterSearcher::Initialize()
-changed Genie version in ClusterSearcher/LoadGenieEventConfig to 1

flemmons commented:
Regarding these [https://github.com/ANNIEsoft/ToolAnalysis/pull/365#issuecomment-4000879256] :
-I've changed the genie version variable to 1, which seems to be used in other toolchains.  Although I will mention that my ToolAnalysis cannot find the files for these, so it can't run with this setting.  I haven't checked the trueQ2 output.  Seeing as how the only thing trueQ2 is used for is direct output to the file in NeutronCheck, I do not feel that this should hold up the PR any further.
-The charge balance constant is a known issue, but it's not as quick a fix as I'd like, since I'm hoping to integrate PMT masking, and make it reactive to the actual number of currently active PMTs.  I will need to pass that information in from other tools.  I will do this in a future update, but seeing as it's not ready to go, I'll request we let the status quo ride this PR into the next one.
-I have removed that line in the new commit.
-I'm not sure you used the correct link.  What memory leak do you mean?

flemmons committed:
Merge branch 'Application' into Application

flemmons committed:
Fixed memory leak caused by double-pointing to the same object
-replaced all instances of fClusterList with fRecoClusters in ClusterSearcher
-changed ClusterSearcher::RecoClusters() to a void function, since it was returning a member pointer to itself
-removed ClusterSearcher::SelectByClusters(), which is not used

marc1uk commented:
Thanks for the updates Franklin. That's got the vast majority of comments resolved, i think we're very nearly there. 

The only one I'm still _somewhat_ reticent to ignore is the '123' constant in the charge balance calculations. 
Part of my concern is that I believed 123 to be a meaningless placeholder value that would give invalid outputs, but perhaps it's not that bad. If this is a hard-coded but **correct** number (based on a quick check of the HV page, ANNIE has ~119 tank PMTs active, so perhaps 123 before bad-channel masking is correct?), then that's a bit better.
 
While applying run-wise bad channel masking would be ideal, I'd still much prefer leaving that as a TODO but still fetching the value from geometry over hard-coding it. It feels like the distinction between "this is a simplification" and "this is sloppy", and it should be trivial to implement.

Aside from that it seems like there's a compilation error flagged by the latest CI build:
```
In member function 'void ClusterSearcher::RecoClusters(std::vector<RecoDigit*>*)':
UserTools/ClusterSearcher/ClusterSearcher.cpp:636:42:
error: no matching function for call to 'RecoCluster::SetDigits(std::vector<RecoDigit>&)' cluster.SetDigits(ClusteredDigits);
                                          ^
In file included from UserTools/ClusterSearcher/ClusterSearcher.h:9,
                 from UserTools/ClusterSearcher/ClusterSearcher.cpp:1:
include/RecoCluster.h:28:15: note: candidate: 'void RecoCluster::SetDigits(std::vector<RecoDigit>*)'
   inline void SetDigits(vector<RecoDigit>* indigits){fDigitList=indigits;}
               ^~~~~~~~~
include/RecoCluster.h:28:15: note:   no known conversion for argument 1 from 'std::vector<RecoDigit>' to 'std::vector<RecoDigit>*'
make[1]: [Makefile:203: UserTools/ClusterSearcher/ClusterSearcher.o] Error 1 (ignored)
```

flemmons committed:
Pointer cleanup in RecoClusterClass

marc1uk commented:
Reply on slack: 
> According to the original phase-II design specifications, 123 is one short of the ideal case with all PMT slots full.  I haven't dug deeply into why yet, but I assume it was instituted after one tube had failed/been removed/not been installed.  There's an empty sconce in the tank, the story of which I don't know.  The 123 was put in by Michael before I got involved, and since it doesn't make much intra-run difference to the measurement (it's largely just a correction so that if all the PMTs see exactly the same charge, the CB becomes 0 for the extreme case), it's been a low priority.  That said, I do plan to fix it when I shuffle in my other neutron identification updates in the next couple of weeks.

Ok, if it's got motivation and precedent, that's good enough for me. :+1:
